server: split health endpoint into health and readiness endpoints

asubiotto opened the pull request:
The health endpoint reverts to returning whether the node liveness is
expired or not. The readiness endpoint returns whether the node is ready
to receive SQL traffic. This is an important distinction needed by load
balancers as an unhealthy node is restarted by force and an unready node
simply does not get traffic sent to it. The new readiness endpoint is
accessed through "_admin/v1/ready"

Addresses #22424
Addresses #22468
Fixes #https://github.com/cockroachdb/cockroachdb-cloudformation/issues/13

Release note (bug fix): Health endpoint has been split into a simpler
health endpoint and a readiness endpoint to integrate better with load
balancers.

cc @bobvawter @nstewart @bdarnell

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/22911)
<!-- Reviewable:end -->

asubiotto commented:
I am planning on cherrypicking this into 2.0 and 1.1

a-robinson commented on pkg/server/admin.go:
```diff
@@ -1040,14 +1041,44 @@ func (s *adminServer) Cluster(
 func (s *adminServer) Health(
 	ctx context.Context, req *serverpb.HealthRequest,
 ) (*serverpb.HealthResponse, error) {
+	isLive, err := s.server.nodeLiveness.IsLive(s.server.NodeID())
```
For what it's worth, comments more clearly explaining the reason for using `IsLive` vs `IsHealthy` here or in node_liveness.go would be nice. To really be sure I had to look at the commit message of bcc6d0b085928fb3cb5faaf2edd074edfa558b2b

asubiotto replied:
Done (added to `node_liveness.go`) let me know if this is what you had in mind.

a-robinson commented on pkg/server/admin.go:
```diff
@@ -1040,14 +1041,44 @@ func (s *adminServer) Cluster(
 ) (*serverpb.HealthResponse, error) {
+	isLive, err := s.server.nodeLiveness.IsLive(s.server.NodeID())
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "node is not live")
```
Why do we return "node is not live" here but the body of the error in the analogous case of the `Ready()` function?

asubiotto replied:
Oversight, thanks for catching! Also changed the `Errorf` calls to just `Error`.

a-robinson commented on pkg/server/admin.go:
```diff
@@ -1040,14 +1041,44 @@ func (s *adminServer) Cluster(
 	}
-	return &serverpb.HealthResponse{}, nil
+
+	// Attempt to execute a SQL query with a timeout.
```
Can you explain for posterity which situations this will catch?

asubiotto replied:
I changed this to observe the `serveMode` on the server, which is set after startup. I think this is the best way to check for readiness since more involved checks are not necessarily what we're going for here. i.e. we're not trying to answer whether or not we can execute SQL queries at the time of the check, we're trying to answer if the server should be able to accept connections.

a-robinson reviewed (approved):
LGTM

bdarnell commented:
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/server/serverpb/admin.proto, line 415 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22911#-L6CNpOppX5bqcPul_sh:-L6CNpOppX5bqcPul_si:bbmhkwi) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb2aa50207375680bd741f9631a197ea8ac8d70e/pkg/server/serverpb/admin.proto#L415)):*
> ```protobuf
> // ReadyRequest inquires whether the addressed node is ready to receive client
> // traffic.
> message ReadyRequest {
> ```

I'd rather add parameters to HealthRequest than an entirely different method. Something like `/health?ready=1`. We may want to add additional options in the future, such as a minimum version or cluster ID (some deployment environments find these options useful to avoid problems caused by back-from-the-dead servers). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22911)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Something I'm confused about: we have another endpoint at `/health` which hits the `Details` endpoint (different from `/_admin/v1/health`). This is kind of confusing and I think we might be using the wrong endpoint in our kubernetes configuration (@nstewart). Does anybody know what's up with that?

---

Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/server/serverpb/admin.proto, line 415 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22911#-L6CNpOppX5bqcPul_sh:-L6INRXrxN0h19ifuAYb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb2aa50207375680bd741f9631a197ea8ac8d70e/pkg/server/serverpb/admin.proto#L415)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I'd rather add parameters to HealthRequest than an entirely different method. Something like `/health?ready=1`. We may want to add additional options in the future, such as a minimum version or cluster ID (some deployment environments find these options useful to avoid problems caused by back-from-the-dead servers). 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22911#-:-L6INSI88vOBDWs7X19j:bc6z772)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
`/health` is the one I thought you were changing here; that's the one we use in our LB configs. I don't know what `/_admin/v1/health` is used for. 

---

Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/server/admin.go, line 1045 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22911#-L6JG8-Mq1pym3AAF8Jr:-L6JG8-Mq1pym3AAF8Js:b-97ulqr) ([raw file](https://github.com/cockroachdb/cockroach/blob/fff4c400fc5a995ebdc2529c3e1a4e8a6bed184b/pkg/server/admin.go#L1045)):*
> ```Go
> 	// This is a simple health check, the client doesn't care whether or not the
> 	// node is ready to receive client traffic.
> 	if !req.Ready {
> ```

"Ready" checks are a superset of regular checks, so I'd like to see this structured as `if !req.Ready { return }` in the middle of this method instead of a completely separate path.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22911#-:-L6JFTq3gtX6Jki5m7Aq:bhqbrlw)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
`/health` only really tells you whether the process is running or not, which isn't particularly useful for indicating whether cockroach is able to serve SQL queries. We should change LB configs to use `/_admin/v1/health?ready=1` once this is in. We use `/_admin/v1/health` today in our DC/OS framework.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22911#-:-L6Mi4AwpfWnGKsFHYDV:bnxt1j2)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
I'm arguing that we should make the endpoint used by load balancers `/health?ready=1`. Doesn't the underscore in `/_admin` mean internal use only?

asubiotto commented:
Review status: 4 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/server/admin.go, line 1045 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22911#-L6JG8-Mq1pym3AAF8Jr:-L6MyjpP4lNf5B-zK5ow:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fff4c400fc5a995ebdc2529c3e1a4e8a6bed184b/pkg/server/admin.go#L1045)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

"Ready" checks are a superset of regular checks, so I'd like to see this structured as `if !req.Ready { return }` in the middle of this method instead of a completely separate path.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22911)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Would we then make `/health` in v2.0 behave like `/_admin/v1/health` behaves in v1.1 then? What's your proposal for moving these endpoints around?

bdarnell commented:
I think `/health` should work as it does today, and `/health?ready=1` should work according to the new changes in this PR. I'm not sure what `/_admin/v1/health` is trying to do or how it is used; I don't see any reason to propagate the stuff that it does to the non-`ready` form of `/health`.

a-robinson commented:
It's very weird from an API design perspective that `/health?ready=1` will have more in common with `/_admin/v1/health` than it will with `/health`.

Just to make sure we're on the same page, the `(*adminServer).Health` method that @asubiotto has been modifying in this PR is the code for `/_admin/v1/health`. The code for `/health` is: https://github.com/cockroachdb/cockroach/blob/e36099f9ee45d53e35e7c2a6b7591fe17307a07f/pkg/server/status.go#L496

bdarnell commented:
> It's very weird from an API design perspective that /health?ready=1 will have more in common with /_admin/v1/health than it will with /health.

True, but it's very weird that we've ended up with two endpoints called "health" that do such different things. I'd rather consolidate all variants of health checking onto `/health` with different options.

a-robinson commented:
Alright, I can live with that. Sounds like @asubiotto has some more work to do then.

Commit:
server: revert /_admin/v1/health behavior

This reverts bcc6d0b085928fb3cb5faaf2edd074edfa558b2b, as the load
balancer endpoint used is /health which hits the Details endpoint.
Additionally, if liveness checks are failed, a node is usually
restarted. We don't want a draining node to be restarted while it is
draining.

Release note: None

Commit:
server: add readiness parameter to /health endpoint

The readiness endpoint (/health?ready=1) returns whether the node is
ready to receive client traffic. This is necessary for load balancers.

Addresses #22424
Addresses #22468
Fixes #cockroachdb-cloudformation/issues/13

Release note (bug fix): readiness endpoint added (/health?ready=1)
for better integration with load balancers.

asubiotto commented:
Added the readiness check to the `Details` endpoint (accessible as `/health?ready=1`). I left `Health()` as is (reverted to its previous behavior). The future of `/_admin/v1/health` is not something I want to tackle in this PR. PTAL. Note that liveness checks will not take into account node liveness.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r4, 2 of 5 files at r5, 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/server/admin.go, line 1043 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/22911#-L6PQ1k6TWS-YeGV_CJv:-L6PQ1k6TWS-YeGV_CJw:b-j55pt8) ([raw file](https://github.com/cockroachdb/cockroach/blob/101db4a8eb3d5845accd7a55d504b28fd12ec122/pkg/server/admin.go#L1043)):*
> ```Go
> 	ctx context.Context, req *serverpb.HealthRequest,
> ) (*serverpb.HealthResponse, error) {
> 	isLive, err := s.server.nodeLiveness.IsLive(s.server.NodeID())
> ```

Is this change from IsHealthy to IsLive still appropriate? This change should at least be called out in the commit message. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22911#-:-L6PQTGCBE7zUHh1F_Je:bnfp4nl)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/server/admin.go, line 1043 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/22911#-L6PQ1k6TWS-YeGV_CJv:-L6SA-lmBAnmQfGMWN66:bgiskk9) ([raw file](https://github.com/cockroachdb/cockroach/blob/101db4a8eb3d5845accd7a55d504b28fd12ec122/pkg/server/admin.go#L1043)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Is this change from IsHealthy to IsLive still appropriate? This change should at least be called out in the commit message. 
</blockquote></details>

The first commit mentions it. I think it is appropriate because the change from `IsLive` to `IsHealthy` was badly informed since I assumed this was the endpoint checked by load balancers.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22911)*
<!-- Sent from Reviewable.io -->
